ca/ca.go

```diff
@@ -396,34 +396,103 @@ func (ca *CertificateAuthorityImpl) IssueCertificate(ctx context.Context, issueR
 	}
 	regID := *issueReq.RegistrationID
 
-	csr, err := x509.ParseCertificateRequest(issueReq.Csr)
+	notAfter, serialBigInt, serialHex, err := ca.generateNotAfterAndSerialNumber()
+	if err != nil {
+		return emptyCert, err
+	}
+
+	certDER, err := ca.issueCertificateOrPrecertificate(ctx, issueReq, notAfter, serialBigInt, "cert")
 	if err != nil {
 		return emptyCert, err
 	}
+
+	cert := core.Certificate{
+		DER: certDER,
+	}
+
+	var ocspResp []byte
+	if features.Enabled(features.GenerateOCSPEarly) {
+		ocspResp, err = ca.GenerateOCSP(ctx, core.OCSPSigningRequest{
+			CertDER: certDER,
+			Status:  "good",
+		})
+		if err != nil {
+			err = berrors.InternalServerError(err.Error())
+			ca.log.AuditInfo(fmt.Sprintf("OCSP Signing failure: serial=[%s] err=[%s]", serialHex, err))
+			// Ignore errors here to avoid orphaning the certificate. The
+			// ocsp-updater will look for certs with a zero ocspLastUpdated
+			// and generate the initial response in this case.
+		}
+	}
+
+	// Store the cert with the certificate authority, if provided
+	_, err = ca.sa.AddCertificate(ctx, certDER, regID, ocspResp)
+	if err != nil {
+		err = berrors.InternalServerError(err.Error())
+		// Note: This log line is parsed by cmd/orphan-finder. If you make any
+		// changes here, you should make sure they are reflected in orphan-finder.
+		ca.log.AuditErr(fmt.Sprintf(
+			"Failed RPC to store at SA, orphaning certificate: serial=[%s] cert=[%s] err=[%v], regID=[%d]",
+			serialHex,
+			hex.EncodeToString(certDER),
+			err,
+			regID,
+		))
+		return emptyCert, err
+	}
+
+	return cert, nil
+}
+
+func (ca *CertificateAuthorityImpl) generateNotAfterAndSerialNumber() (time.Time, *big.Int, string, error) {
+	notAfter := ca.clk.Now().Add(ca.validityPeriod)
+
+	// We want 136 bits of random number, plus an 8-bit instance id prefix.
+	const randBits = 136
+	serialBytes := make([]byte, randBits/8+1)
+	serialBytes[0] = byte(ca.prefix)
+	_, err := rand.Read(serialBytes[1:])
+	if err != nil {
+		err = berrors.InternalServerError("failed to generate serial: %s", err)
+		ca.log.AuditErr(fmt.Sprintf("Serial randomness failed, err=[%v]", err))
+		return time.Time{}, nil, "", err
+	}
+	serialBigInt := big.NewInt(0)
+	serialBigInt = serialBigInt.SetBytes(serialBytes)
+	serialHex := core.SerialToString(serialBigInt)
+
+	return notAfter, serialBigInt, serialHex, nil
```

Comment:
@rolandshoemaker Thanks for the feedback.

> (given core.SerialToString is pretty cheap I'm somewhat in favor of just returning the big.Int and letting both IssueCertificate and issueCertificateOrPrecertificate call it themselves).

I took your suggestion and improved this in commit d2e19b3: “CA: Don't compute serialHex in generateNotAfterAndSerialNumber().”

Thanks again!